### PR TITLE
Add ability to merge hit objects in osu! editor to create sliders

### PR DESCRIPTION
**Pull request by OliBomby**
I added a feature to the osu editor which lets you merge multiple objects into one big slider. You can use it by selecting multiple sliders and then pressing the 'Merge selection' in the context menu or pressing Ctrl+Shift+M.

https://user-images.githubusercontent.com/17460441/184664631-decf2530-d119-4198-94ac-250091470893.mp4

### CODE REVIEW

File: osu.Game.Rulesets.Osu.Tests/Editor/TestSceneObjectMerging.cs
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/Editor/TestSceneObjectMerging.cs:4]: Should probably remove this, we avoid adding it for new files.

Comment (bdach) [reply osu.Game.Rulesets.Osu.Tests/Editor/TestSceneObjectMerging.cs:1]: This comment concerns `#nullable disable` and not the copyright notice. [b5e5411](https://github.com/ppy/osu/pull/19782/commits/b5e54113485d5c511bc179a67a2f0794cd78ec55) is wrong.

@OliBomby Please do not resolve review conversations as the author of the code. In the way we work, the reviewer (or another maintainer) is responsible for this.

Comment (OliBomby) [reply osu.Game.Rulesets.Osu.Tests/Editor/TestSceneObjectMerging.cs:1]: oh sorry, i blundered

**Comment by peppy** [comment]
Quite solid in general. I've made a few menial fixes (please read through each commit to understand what/why I changed).

Remaining points of concern:

- I think it would be good to have a test covering the case where objects are selected that can't be merged (and checking the merge doesn't happen)
- The first test brings up an interesting consideration: it merged two objects separated by quite some time, but after merge the second object is moved closer to the first. Is this the expected behaviour? We have per-object SV so it's feasible to keep the duration intact as well by adjusting the SV when merging.

To show what I mean:

https://user-images.githubusercontent.com/191335/184814895-7d9e68c4-25ca-4580-8622-28404b8b204e.mp4

**Comment by bdach** [comment]
> The first test brings up an interesting consideration: it merged two objects separated by quite some time, but after merge the second object is moved closer to the first. Is this the expected behaviour? We have per-object SV so it's feasible to keep the duration intact as well by adjusting the SV when merging.

Wouldn't implementing this require adjusting the SV only on the segment of the merged slider that's connecting the parts being merged, rather than the entire slider? I don't think we have a facility to do that right now?

Also trying to preserve the "time dimension" comes with extra complications, such as, you'd have to disallow performing the merge operation if the objects overlap time-wise in any way, since there's no way to merge them without... well, time travel, I guess?

**Comment by OliBomby** [comment]
@bdach I don't think he meant preserving the duration of all parts of the merged slider. A much more feasible thing to do is to preserve the latest end time of the selected objects. This would be the same as first merging the objects with the current implementation and then shift-click dragging the end to the time of the last hitobject. The current implementation just preserves the SV of the first object.
If I were to choose to preserve the end time then it might be problematic when you merge hitcircles which are all on the same tick, resulting in a slider which should have 0 duration, so I'm not sure what is the better UX. I'm tempted to stick with the current implementation because that's what Slider Merger has been using all this time and no-one ever complained about this.

**Comment by bdach** [comment]
> A much more feasible thing to do is to preserve the latest end time of the selected objects.

Sure, that works. I guess I just wasn't sure what was meant above.

> it might be problematic when you merge hitcircles which are all on the same tick, resulting in a slider which should have 0 duration

You wouldn't allow this operation to proceed in the first place, I don't think? (see my note on "disallowing merge on overlapping objects" above) We've been cracking down on zero-length/zero-duration sliders in lazer - it should be impossible to get the lazer editor to produce one.

**Comment by peppy** [comment]
I think we can add more options at a later point if people ask for it. Let's go with this as-is for now.

File: osu.Game.Rulesets.Osu/Edit/OsuSelectionHandler.cs
Comment (peppy) [inline osu.Game.Rulesets.Osu/Edit/OsuSelectionHandler.cs:338]: Kinda arbitrary at this point... but we can do with it for now. I preferred the inclusive specification because it avoids future hitobjects potentially becoming mergeable but causing an issue.

**Review by peppy — APPROVED** [review]

**Pull request merged**
